### PR TITLE
gpu stanmathcl [WIP]

### PR DESCRIPTION
**Pull request by SteveBronder**
#### Submission Checklist

- [ ] Run unit tests: `./runTests.py test/unit`
- [ ] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Copyright and Licensing

Copyright:

Faculty of Computer and Information Science, University of Ljubljana.
Steve Bronder

Both parties agree to submit the Pull Request under the BSD 3 license

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by bob-carpenter** [comment]
@SteveBronder  Thanks!

Before we can merge, we need clear statements of copyright holder and licensing under BSD-3.  Presumably that's you personally, but you need to add that to the pull request (there's a spot at the bottom of the form).

**Comment by SteveBronder** [comment]
@bob-carpenter  np!

The underlying algorithms are from @rok-cesnovar and his team. I believe we agreed it can all go under the BSD-3 license

I'm going to look this over in the next few days, merge this into the gpu branch inside of the stan math repo, then we can start the actual review

**Comment by rok-cesnovar** [comment]
Yes, we agreed on BSD-3. 

For the underlying algorithm the copyright holder is Faculty of Computer and Information Science, University of Ljubljana.

Thanks, Steve!

**Comment by rok-cesnovar** [comment]
@SteveBronder I made some minor changes, all tests still pass. 

Tomorrow I will do some performance analysis and post what to expect from all this.

**Comment by SteveBronder** [comment]
@rok-cesnovar awesome!

**Comment by rok-cesnovar** [comment]
First the results for the /prim/mat/fun cholesky_decompose_gpu compared to the CPU version. 

- the GPU version is faster than CPU/Eigen version at matrices with M=200
- at M~1000 the speedup is ~8-10
- at M~2500 the speedup is ~14
- at M~4000 the speedup is ~20

Timing and speedup plot: 
time_chol_prim

speedup

I plotted the GPU times without copying data to and from the GPU, so you could see what could be achieved if we found a way of leaving the data on the GPU for the next iteration/function if the next iteration/function is also done on the GPU. As you can see, speedups of close to 100 could be achieved this way. 

Here are the execution times: 
[https://www.dropbox.com/s/akndbsojte2z38t/timing_gpu_cholesky_prim.txt?dl=0](url)

The /rev/mat/fun cholesky_decompose_gpu timings take a bit more time, will post those once I have them.

**Comment by SteveBronder** [comment]
That's awesome! Can you post the code you used to run the speed test?

**Comment by rok-cesnovar** [comment]
@SteveBronder sure, I will make a repo and will link it here. But it wont be until tomorrow, it is getting late in the CEST time zone :)

Here are the results for the /rev/mat/fun cholesky_decompose_gpu compared to the block CPU version that is currently used in rev/mat/fun cholesky_decompose_gpu. The CPU version of the algorithm we parallelized is slower than the block CPU. We have already looked into making a GPU version of the block algorithm, but for now this should do.

Summary:
- the GPU version is faster for matrices over M=50
- at M~1000 the speedup is ~10
- at M~3000 is ~20

Without copying the data to and from the GPU the speedups are around 50 for M larger than 1200.

rev_time

rev_speedup

**Comment by SteveBronder** [comment]
@rok-cesnovar I updated the branch so it passes cpplint. I think our next steps are:
[ ] Documentation 
[ ] Build tests for all the gpu functions
[ ] Figure out the best way for users to touch this (this is more of a Stan team thing which we have spoken about previously)
[ ] Discuss with the team the best way to review this
[ ] Request reviews

@syclik I'm unsure where some of the files should go and whether some file should be broken up into multiple files.

EX1: [check_gpu.hpp](https://github.com/rok-cesnovar/math/blob/gpu_stanmathcl/stan/math/prim/mat/err/check_gpu.hpp)
EX2: [basic_matrix_gpu.hpp](https://github.com/rok-cesnovar/math/blob/gpu_stanmathcl/stan/math/prim/mat/fun/basic_matrix_gpu.hpp)
EX3: [kern_gpu] file (https://github.com/rok-cesnovar/math/tree/gpu_stanmathcl/stan/math/prim/mat/kern_gpu)

**Comment by rok-cesnovar** [comment]
Agreed. I will start adding doxygen comments. 

Regarding the tests: for some of them we can just reuse the cpu tests, but functions like transpose and copying of lower triangular to the upper triangular probably do not have the equivalent.

**Comment by SteveBronder** [comment]
Awesome I'll start building out the tests

**Comment by SteveBronder** [comment]
@seantalts what is the level of coverage we need for the tests? For instance, does the [`zeros()` function](https://github.com/rok-cesnovar/math/blob/3dc3f5dfc28966a139f6cc81f37f60003cb9d623/stan/math/prim/mat/fun/basic_matrix_gpu.hpp#L89) need a unit test?

**Comment by seantalts** [comment]
This is a nuanced question. I think my own framework for thinking about it is that one needs tests that make sure all of the functionality you care about (or that someone might reasonably expect from the API) works correctly under a variety of inputs. So you could test that `zeros` works directly (easy enough to add a quick one) or you could test that functions that rely on it working correctly also work. @syclik might have more advice or a more stringent opinion here.

**Comment by syclik** [comment]
I agree with @seantalts that it's a nuanced question.

My soft requirement (or strong opinion) is that every function / method gets at least one test. I'm not looking for branch coverage or anything like that (although if the function is complicated, you should). Adding a simple test really cuts down on maintenance cost, even if it doesn't seem like it's doing much. Think about coming back to this function months from now. Just having the instantiation of `A` and `part` makes it really easy to track down a bug if you're triggering that `cl::Error`. Or it makes it easy to refactor the code because there's already some scaffolding in place to add more tests if that's what's required. Just think about how quick it would be for you to write the test now vs. coming back to this and having to figure out how to write the test with fresh eyes.

Of course, you also run fix design problems. If you can't write tests easily, then maybe there's a better design that should be employed.

**Comment by SteveBronder** [comment]
@syclik that's reasonable

**Comment by SteveBronder** [comment]
Functions that need tests
- [ ] matrix_gpu functions
- [ ] Compile kernels in ocl_gpu
- [ ] check_opencl_error
- [ ]  transpose (both src -> dst and in-place
- [ ]  zeros
- [ ]  identity
- [ ] copy_triangular
- [ ] copy_submatrix
- [ ] copy_triangular_transposed
- [ ] add
- [ ] subtract
- [ ] lower_triangular_inverse
- [ ] [diagonal_multiply](https://github.com/rok-cesnovar/math/blob/0768acb6c652ff718a386ab58647e3c33e72dcfb/stan/math/prim/mat/fun/multiply_gpu.hpp#L25)
- [ ] [multiply](https://github.com/rok-cesnovar/math/blob/0768acb6c652ff718a386ab58647e3c33e72dcfb/stan/math/prim/mat/fun/multiply_gpu.hpp#L55)

**Comment by rok-cesnovar** [comment]
What should we do in cases where the operation is not thread safe, for instance multiply(A,A,A) which is essentialy A <- A*A. Should we throw an exception or cover this cases with temporary created buffers?

**Comment by seantalts** [comment]
You mean that the algorithm you are using for multiplication can’t be used in-place, right? Is there one that can be? If not I’d say use a temporary buffer.

**Comment by bob-carpenter** [comment]
@rok-cesnovar  When would `multiply(A, A, A)` come up?  It's not something that can be expressed as a Stan function because all the Stan functions copy on output.

**Comment by rok-cesnovar** [comment]
Yeah, the GPU function as it is written now, uses the third parameter as the result matrix. multiply(A,B,C) is therefore C<- A*B. Rewriting this to copy on output also resolves the  A<-A*A issue. 

Thanks!

**Comment by rok-cesnovar** [comment]
I have a question regarding the cholesky_decompose_test, specifically the test_chol_mult and the x_c matrix at line 68 here: https://github.com/stan-dev/math/blob/develop/test/unit/math/rev/mat/fun/cholesky_decompose_test.cpp

It looks to me as if parts of the matrix is left uninitialized and that can cause some issues from time to time. I noticed some weird behaivor when running the tests as the initial input to the cholesky_decompose sometimes had NaN or Inf values. This is before any of the new added code is called. 

I then ran a few tests on the develop branch (no GPU code at all) and noticed something weird and would just like to double check if this is the expected behavior or is this a possible bug. 

These are some printed out examples of x_c from the above mentioned test. In all cases I ran test_chol_mult(25, 1e-08); 


I first ran all the tests in cholesky_decompose_test.cpp and got this matrix [test_1.txt](https://github.com/stan-dev/math/files/1467947/test_1.txt)
The input was the same on each run. 

I then commented out all the tests except for the the mat_cholesky_1st_deriv_large_gradients. I got this for the input matrix 
[test_2.txt](https://github.com/stan-dev/math/files/1467957/test_2.txt)

I then swapped the test_gp_grad(1000, 1e-08) and test_gradient(36, 1e-08); tests and got this input matrix 
[test_4.txt](https://github.com/stan-dev/math/files/1467968/test_4.txt)

Basically every time I change any parts of the tests or underlying stan::math code, part of the input changes. In some extreme cases I got NaN/Inf values in the bottom right part of the input matrix.
However, in all cases, the upper left part of the matrix is the same. 

Am I missing something or is this a bug in the test? Thank you for time!

**Comment by seantalts** [comment]
Oh thank god. You're right, that test fails from time to time (pretty rarely but I've seen it a few times this year). Want to make a separate PR to fix that?

**Comment by rok-cesnovar** [comment]
Sure, but I am not sure I have enough knowledge of the actual tests to know how to fix this.

**Comment by seantalts** [comment]
Which part is left uninitialized? Are you saying `cov_matrix_constrain` doesn't initialize all of the matrix?

**Comment by rok-cesnovar** [comment]
I think the problem is in the input x of the cov_matrix_constrain. It looks that x is uninitialized.

**Comment by seantalts** [comment]
It seems like x is passed in, right? So it should theoretically be initialized by `stan::math::gradient` and `finite_diff_gradient` on lines 288 and 294?

I do think there's something that isn't initialized...

**Comment by rok-cesnovar** [comment]
I think I found it. 

The problem is in the for loop at line 279.
`test_vec.size()` is smaller than `vec_size` 

I changed the loop to this:
```
for (int i = 0; i < vec_size; ++i) {
    test_vals(i) = i % 10 / 100.0;
    if (i < test_vec.size())
      test_vec(i) = stan::math::normal_rng(0.0, 0.1, rng);    
  }
```
This fixes the problem.

**Comment by SteveBronder** [comment]
@rok-cesnovar nice we are passing all the tests!

I have two write three more tests for the underlying cholesky decomposition kernels and fix up the doxygen so we pass the jenkins build. But looks like we are good! I'll do this tmrw after work

**Comment by rok-cesnovar** [comment]
The underlying cholesky decomposition was rewritten a bit so that it reuses all the other GPU functions, except for a short kernel that calculates the cholesky for smaller blocks. So that one is the only that needs a separate test now. I will remove the remaining ones.

Would it be of any value for the review process or in general if I made a PDF with a short explanation of the idea behind the cholesky GPU implementation? That one and the inversion are more complex, others are either standard GPU implementations (matrix multiply) or simple (transpose, zeros, identity,...).

**Comment by syclik** [comment]
Yes to the PDF! Doc is helpful; especially motivation for design. Aim it at people that are in between novice and expert and it'll come in handy for the future too. 

Thanks!

**Comment by SteveBronder** [comment]
Jenkins, ok to test.

File: stan/math/prim/arr/err/check_opencl.hpp [hunk 1]
```diff
@@ -0,0 +1,304 @@
+#ifndef STAN_MATH_PRIM_ARR_ERR_CHECK_OPENCL_HPP_
+#define STAN_MATH_PRIM_ARR_ERR_CHECK_OPENCL_HPP_
+
+#include <CL/cl.hpp>
+#include <iostream>
+#include <string>
+
+
+/** @file stan/math/prim/arr/err/check_opencl.hpp
```
Comment (bob-carpenter) [inline stan/math/prim/arr/err/check_opencl.hpp]: This can be simplified with a throw function to avoid assignments and breaks.

I assume throwing on 0 was a mistake.

```
void throw_openCL(const std::string& function, const char* msg) {
  throw std::domain_error(function + ": The OpenCL application ended with the error: " + msg);
}
...

  switch(e.err()) {
    case 0:
      return;  // ??? should this throw instead ???
    case 1:
      throw_openCL("CL_DEVICE_NOT_FOUND");
...
```

Given that they're going to be Stan errors, we could also add some explanatory text.

File: stan/math/prim/arr/fun/matrix_gpu.hpp [hunk 2]
```diff
@@ -25,169 +20,195 @@ enum copy_transposed_triangular {LOWER_TO_UPPER_TRIANGULAR = 0,
...
+         * only allocates the buffer on the GPU.
+         * 
+         * @param rows number of matrix rows
+         * @param cols number of matrix columns
+         * 
+         * @throw <code>std::invalid_argument</code> if the
+         * matrices do not have matching dimensions
+         * 
+         */
+        matrix_gpu(int rows,  int cols)
```
Comment (bob-carpenter) [inline stan/math/prim/arr/fun/matrix_gpu.hpp]: Would it be easier to wrap the OpenCL with try/catch directly, like `get_context()` and `cl::buffer()`?  Then it simplifies all the matrix code and scopes the try/catch appropriately.

File: stan/math/prim/arr/fun/matrix_gpu.hpp [hunk 3]
```diff
@@ -25,169 +20,195 @@ enum copy_transposed_triangular {LOWER_TO_UPPER_TRIANGULAR = 0,
...
           try {
-            cl::Context ctx = stan::math::get_context();
-            oclBuffer = cl::Buffer(ctx, CL_MEM_READ_WRITE,
-             sizeof(double) * rows * cols);
-            this->rows = rows;
-            this->cols = cols;
-          } catch (cl::Error& e) {
-            check_ocl_error(e);
+            cl::Context& ctx = get_context();
+            if (rows_ > 0 && cols_ > 0) {
```
Comment (bob-carpenter) [inline stan/math/prim/arr/fun/matrix_gpu.hpp]: Does this leave `oclBuffer` undefined if `rows_ > 0` or `cols_ > 0`?  Should this just throw an exception instead?  Should this check go before getting a context then?

File: stan/math/prim/arr/fun/matrix_gpu.hpp [hunk 4]
```diff
@@ -25,169 +20,195 @@ enum copy_transposed_triangular {LOWER_TO_UPPER_TRIANGULAR = 0,
...
-            queue.enqueueWriteBuffer(oclBuffer, CL_TRUE, 0,
-             sizeof(double) * A.rows() * A.cols(), Atemp);
-            delete[] Atemp;
-          } catch (cl::Error& e) {
-            check_ocl_error(e);
+            cl::Context& ctx = get_context();
+            cl::CommandQueue& queue = get_queue();
+            rows_ = A.rows();
+            cols_ = A.cols();
+            if (rows_ > 0 && cols_ > 0) {
```
Comment (bob-carpenter) [inline stan/math/prim/arr/fun/matrix_gpu.hpp]: This can just test `A.size() > 0`.

File: stan/math/prim/arr/fun/matrix_gpu.hpp [hunk 5]
```diff
@@ -25,169 +20,195 @@ enum copy_transposed_triangular {LOWER_TO_UPPER_TRIANGULAR = 0,
...
+     * The matrix data is stored in the oclBuffer.     
+     * 
+     */
     class matrix_gpu {
       private:
-
         cl::Buffer oclBuffer;
 
       public:
+        int rows_;
```
Comment (bob-carpenter) [inline stan/math/prim/arr/fun/matrix_gpu.hpp]: `rows_` and `cols_` should also be private if you have a `rows()` and `cols()` accessor.  Does anything need to set them?  Can they be constant?

File: stan/math/prim/arr/fun/matrix_gpu.hpp [hunk 6]
```diff
@@ -25,169 +20,195 @@ enum copy_transposed_triangular {LOWER_TO_UPPER_TRIANGULAR = 0,
...
+     * @throw <code>std::invalid_argument</code> if the 
+     * matrices do not have matching dimensions
+     * 
+     */
+    void copy(matrix_gpu& src,  matrix_gpu& dst) { // NOLINT
+      check_size_match("copy (GPU -> GPU)",
+        "src.rows()", src.rows(), "dst.rows()", dst.rows());
+      check_size_match("copy (GPU -> GPU)",
+        "src.cols()", src.cols(), "dst.cols()", dst.cols());
+      cl::Kernel kernel = get_kernel("copy");
```
Comment (bob-carpenter) [inline stan/math/prim/arr/fun/matrix_gpu.hpp]: Is this outside of the try/catch because it can't throw?

File: stan/math/prim/mat/err/check_gpu.hpp [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+     * @param name Variable name (for error messages)
+     * @param y Matrix on the GPU to test
+     *
+     * @throw <code>std::invalid_argument</code> if the matrix
+     *    is not square
+     */
+    inline void
+    check_square(const std::string& function,
+                 const std::string& name,
+                 matrix_gpu& y) {
```
Comment (bob-carpenter) [inline stan/math/prim/mat/err/check_gpu.hpp]: `matrix_gpu` should be `const`---this shouldn't be modifying the value.

File: stan/math/prim/arr/fun/matrix_gpu.hpp [hunk 8]
```diff
@@ -25,169 +20,195 @@ enum copy_transposed_triangular {LOWER_TO_UPPER_TRIANGULAR = 0,
...
+     */
     class matrix_gpu {
       private:
-
         cl::Buffer oclBuffer;
 
       public:
+        int rows_;
+        int cols_;
+        int rows() {
```
Comment (bob-carpenter) [inline stan/math/prim/arr/fun/matrix_gpu.hpp]: These methods should be declared `const` so that they can be used on `const` arguments (this is required for a fix mentioned below to make a `matrix_gpu` argument constant).  E.g.,

```
int rows() const { return rows_; }
```

File: stan/math/prim/mat/err/check_gpu.hpp [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+     *
+     * @param function Function name (for error messages)
+     * @param name Variable name (for error messages)
+     * @param y Matrix to test
+     *
+     * @throw <code>std::domain_error</code> if
+     *    any element of the matrix is <code>NaN</code>.
+     */
+    inline void
+    check_positive_definite_gpu(const std::string& function,
```
Comment (bob-carpenter) [inline stan/math/prim/mat/err/check_gpu.hpp]: It seems like there might be somewhere else where all this work is done so it isn't duplicated inside `check_positive_definite`.  It can be overloaded instead of using the suffix `_gpu`.

File: stan/math/prim/mat/err/check_gpu.hpp [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+     * @param function Function name (for error messages)
+     * @param name1 Variable name for the first matrix (for error messages)
+     * @param y1 First matrix
+     * @param name2 Variable name for the second matrix (for error messages)
+     * @param y2 Second matrix
+     *
+     * @throw <code>std::invalid_argument</code>
+     * if the dimensions of the matrices do not match
+     */
+    inline void check_matching_dims(const std::string& function,
```
Comment (bob-carpenter) [inline stan/math/prim/mat/err/check_gpu.hpp]: Using `_gpu` as a suffix should be consistent---either do it everywhere or nowhere.

File: stan/math/prim/mat/fun/cholesky_decompose_gpu.hpp [hunk 11]
```diff
@@ -20,117 +22,179 @@
...
+        zeros(A, UPPER);
+      } catch (const cl::Error& e) {
+        check_ocl_error("cholesky_decompose", e);
+      }
+      copy_triangular_transposed(A, LOWER_TO_UPPER_TRIANGULAR);
+      check_positive_definite_gpu("cholesky_decompose_gpu",
+        "Matrix m", A);
+      zeros(A, UPPER);
+      matrix_gpu B(A.rows(), A.cols());
+      copy(A, B);
```
Comment (bob-carpenter) [inline stan/math/prim/mat/fun/cholesky_decompose_gpu.hpp]: This might be clearer as a copy constructor

```
matrix_gpu B(A);
```

so you'd define

```
matrix _gpu(const matrix_gpu& A);
```
if you need something other than default behavior.

File: stan/math/prim/arr/err/check_opencl.hpp (hunk 1 above)
Comment (rok-cesnovar) [reply stan/math/prim/arr/err/check_opencl.hpp]: You are right, the throw on 0 was a mistake. The throw function was added as suggested. Thank you!

File: stan/math/prim/arr/fun/matrix_gpu.hpp (hunk 5 above)
Comment (rok-cesnovar) [reply stan/math/prim/arr/fun/matrix_gpu.hpp]: The previous version of the transpose function, these values were modified. Since the change of the transpose to copy on output, this is no longer needed, but was not changed to const. Fixed now. Thank you!

File: stan/math/prim/arr/fun/matrix_gpu.hpp (hunk 8 above)
Comment (rok-cesnovar) [reply stan/math/prim/arr/fun/matrix_gpu.hpp]: Fixed, explained above.

File: stan/math/prim/arr/fun/matrix_gpu.hpp (hunk 2 above)
Comment (rok-cesnovar) [reply stan/math/prim/arr/fun/matrix_gpu.hpp]: The get_context/get_queue just return the references if the context/queue was already initialized. Otherwise, init() is called, which has its own try/catch. I modified this to only wrap the OpenCL part.

File: stan/math/prim/arr/fun/matrix_gpu.hpp (hunk 3 above)
Comment (rok-cesnovar) [reply stan/math/prim/arr/fun/matrix_gpu.hpp]: The stan math [tests](https://github.com/stan-dev/math/blob/b9844645a369f3187342f14abe3ded328c423583/test/unit/math/prim/mat/fun/multiply_test.cpp#L40) do not expect an exception when a matrix/vector/row_vector of size 0 is used.

The OpenCL cl::Buffer constructor throws an exception if the size of Buffer is 0, so this check was added. I moved it before the get_context, which is obviously unnecessary, if size() is 0.

File: stan/math/prim/arr/fun/matrix_gpu.hpp (hunk 4 above)
Comment (rok-cesnovar) [reply stan/math/prim/arr/fun/matrix_gpu.hpp]: Fixed. Thank you!

File: stan/math/prim/arr/fun/matrix_gpu.hpp (hunk 6 above)
Comment (rok-cesnovar) [reply stan/math/prim/arr/fun/matrix_gpu.hpp]: The try/catch scopes scopes were a bit too big (included stuff that either did not throw or has its own try/catch for OpenCL code. This was addressed in the latest commit.

File: stan/math/prim/mat/err/check_gpu.hpp (hunk 7 above)
Comment (rok-cesnovar) [reply stan/math/prim/mat/err/check_gpu.hpp]: `const` was added to all the check functions, we are checking other non-check function that could also have input as `const`

File: stan/math/prim/mat/err/check_gpu.hpp (hunk 10 above)
Comment (rok-cesnovar) [reply stan/math/prim/mat/err/check_gpu.hpp]: The `_gpu` suffix was removed. The only remaining function with the `_gpu` suffix is the cholesky_decompose_gpu, which can not be overloaded as it expects the same arguments as the existent cholesky_decompose function

**Comment by SteveBronder** [comment]
@rok-cesnovar fyi I tried to rebase to master last night and messed something up. Will fix tonight

**Comment by rok-cesnovar** [comment]
Deal. We are working on the documentation that will explain the more complex parts of this implementations which will hopefully help the reviewers.

**Comment by SteveBronder** [comment]
Jenkins, ok to test

**Comment by syclik** [comment]
@SteveBronder, no need to say "Jenkins, ok to test" -- once it's been ok'ed, every push on the branch will be tested.

**Comment by SteveBronder** [comment]
@syclik Apologies, I'm getting two errors on jenkins that I'm confused about
1.
```bash
duplicate symbol __ZN4stan4math10get_kernelENSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE in:
    test/unit/multiple_translation_units1.o
    test/unit/multiple_translation_units2.o
```
2.
```
Recording test results
ERROR: Step ‘Publish JUnit test result report’ failed: No test report files were found. Configuration error?
Finished: FAILURE
```

How do I deal with the multiple translation units and did I do something wrong with the Jenkins Config file? Looking at my changed files it does not look like I touched the Jenkins file.

**Comment by seantalts** [comment]
I'm not positive but try declaring your functions (esp seemingly
"get_kernel") as `inline`? I think that's the main trick we use to allow
our header-only library to be used in multiple translation units. Search
around the github issues and stuff for that phrase as well

**Comment by SteveBronder** [comment]
Jenkins is currently failing with
@syclik @bob-carpenter 
```
g++-4.9 -Wall -I . -isystem lib/eigen_3.3.3 -isystem lib/boost_1.64.0 -isystem lib/cvodes_2.9.0/include -isystem lib/opencl22 -std=c++1y -DBOOST_RESULT_OF_USE_TR1 -DBOOST_NO_DECLTYPE -DBOOST_DISABLE_ASSERTS -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION -Wno-unused-function -Wno-uninitialized -DSTAN_GPU -DNO_FPRINTF_OUTPUT -pipe  -c -O0 -include stan/math/fwd/scal/fun/acosh.hpp test/dummy.cpp -o /dev/null
In file included from ./stan/math/prim/scal/fun/acosh.hpp:4:0,
                 from ./stan/math/prim/mat/fun/acosh.hpp:5,
                 from ./stan/math/prim/mat.hpp:49,
                 from ./stan/math/fwd/mat.hpp:9,
                 from <command-line>:0:
./stan/math/prim/scal/fun/constants.hpp: At global scope:
./stan/math/prim/scal/fun/constants.hpp:161:18: warning: ‘stan::math::NEG_TWO_OVER_SQRT_PI’ defined but not used [-Wunused-variable]
     const double NEG_TWO_OVER_SQRT_PI = -TWO_OVER_SQRT_PI;
```

Did I rebase wrong and this turned up somewhere?

**Comment by syclik** [comment]
@SteveBronder, want to take care of the error first? http://d1m1s1b1.stat.columbia.edu:8080/job/Math%20Pull%20Request%20-%20Tests%20-%20Header/949/console
```
g++-4.9 -Wall -I . -isystem lib/eigen_3.3.3 -isystem lib/boost_1.64.0 -isystem lib/cvodes_2.9.0/include -isystem lib/opencl22 -std=c++1y -DBOOST_RESULT_OF_USE_TR1 -DBOOST_NO_DECLTYPE -DBOOST_DISABLE_ASSERTS -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION -Wno-unused-function -Wno-uninitialized -DSTAN_GPU -DNO_FPRINTF_OUTPUT -pipe  -c -O0 -include stan/math/fwd/scal/fun/is_inf.hpp test/dummy.cpp -o /dev/null
In file included from ./stan/math/prim/mat/fun/ocl_gpu.hpp:6:0,
                 from ./stan/math/prim/mat/fun/cholesky_decompose_gpu.hpp:4,
                 from ./stan/math/prim/mat.hpp:247,
                 from ./stan/math/rev/mat.hpp:12,
                 from ./stan/math.hpp:4,
                 from <command-line>:0:
./stan/math/prim/arr/err/check_opencl.hpp: In function ‘void stan::math::throw_openCL(const char*, const char*)’:
./stan/math/prim/arr/err/check_opencl.hpp:29:9: error: invalid operands of types ‘const char*’ and ‘const char [48]’ to binary ‘operator+’
         ": The OpenCL application ended with the error: " + msg);
         ^
```

I'm not sure how the warnings aren't triggered in other PRs. I'll have to dig a bit.

**Comment by SteveBronder** [comment]
@rok-cesnovar On Jenkins we are still getting that OpenCL error for an invalid work group size

http://d1m1s1b1.stat.columbia.edu:8080/job/Math%20Pull%20Request%20-%20Tests%20-%20Unit/949/console

I'm looking into whether we can use `CL_DEVICE_MAX_WORK_GROUP_SIZE` from [this](https://stackoverflow.com/questions/5557164/causes-for-cl-invalid-work-group-size] stackoverflow answer to allocate the work group size on a per machine basis.

**Comment by rok-cesnovar** [comment]
Yeah, we can use that. It seems that the CPU on which Jenkins runs this only supports 64 work-items per group. I will fix this tonight, thanks.

**Comment by SteveBronder** [comment]
@syclik How do I interpret the current errors in Jenkins?

http://d1m1s1b1.stat.columbia.edu:8080/job/Math%20Pipeline/job/PR-655/2/

Looking at the [console output](http://d1m1s1b1.stat.columbia.edu:8080/job/Math%20Pipeline/job/PR-655/2/console) and [content of the test](http://d1m1s1b1.stat.columbia.edu:8080/job/Math%20Pipeline/job/PR-655/2/warnings2Result/new/package.-93036363/source.5155416026359294127/#1)(I think?) I'm not really sure how to interpret the errors

**Comment by bob-carpenter** [comment]
I find the console output easiest to digest:

http://d1m1s1b1.stat.columbia.edu:8080/job/Math%20Pipeline/job/PR-655/2/console

Looks like a bunch of failed unit tests for some reason that seems unrelated
to anything you'd be doing with GPUs.

**Comment by SteveBronder** [comment]
Thanks I'll look at the console in the future. Should I have Jenkins retest this or should someone else look into what is going on?

ftr, I am expecting an error for invalid work group sizes ala [this](http://d1m1s1b1.stat.columbia.edu:8080/job/Math%20Pull%20Request%20-%20Tests%20-%20Unit/949/console) console output from a previous run.

**Comment by seantalts** [comment]
I also like the pipeline view: http://d1m1s1b1.stat.columbia.edu:8080/job/Math%20Pipeline/job/PR-655/2/flowGraphTable/

In [the Unit phase ](http://d1m1s1b1.stat.columbia.edu:8080/job/Math%20Pipeline/job/PR-655/2/execution/node/88/log/) I see a lot of this message:

```
clang: : error: : -framework OpenCL: 'linker' input unused
```

**Comment by syclik** [comment]
Just recapping status: we are waiting on other pull requests.
- #720
- #721

**Comment by syclik** [comment]
@SteveBronder, mind if I close this PR for now? We can reopen once we have the other pieces in. (Just trying to keep a handle on open Math PRs.)

**Comment by SteveBronder** [comment]
Totally, will do now

**Comment by seantalts** [comment]
It'd be nice if there was a way to keep the discussion for the GPU (or MPI) topic in fewer places, though I suppose at some point that discussion needs to be summarized and documented with the code or new PR. I'm always surprised when I'm following a PR and someone mentions that there's a corresponding thread on discourse with some of the same and some new information.

**Comment by syclik** [comment]
+1.

**Comment by syclik** [comment]
@seantalts: if you want to be proactive about it, want to come up with a draft of guidelines that we can discuss / encourage others to follow?

It used to be: discuss the goals and design on stan-dev email list, put down a concrete issue on GitHub with clarification on the issue (but no real discussion), then a pull request was code-level comments. But that was just a social norm and never written down.

As we've gotten new contributors and different tooling (discourse + GitHub adding templates, projects, and making it easier to merge across forks), we haven't been able to encourage putting the discussions in the right place. Putting up guidelines that we can point people to would be a good start.

**Comment by syclik** [comment]
If that's enough of a description, I can just put that on a wiki and update the Guidelines (in the repo).

**Comment by bob-carpenter** [comment]
What do you recommend to keep discussion in one place?  Start a thread on discourse to discuss one-location discussions and link from here? :-)

**Comment by SteveBronder** [comment]
From the outside perspective it's also hard to decide where to post about things. ie I made this discourse thread [here](http://discourse.mc-stan.org/t/catching-gpu-errors-making-new-error-codes-or/3384/1) about system errors, but thinking about it now this should have been posted on the PR.

I think it could be something like
1. Proposal on discourse
2. File detailed proposal as a github issue (possibly with MVP included)
3. Make a fork/branch with first draft and submit PR
4. Discuss code and CI issues on PR and design issues on discourse (?)
5. Iterate until ready for review
6. ???
7. Merged!

**Comment by seantalts** [comment]
So there could be a few ways slice and dice what goes where. First though, I think maybe it could be helpful to have a mechanical rule that is always followed (and doesn't require judgement calls) we could go a long way:

1. If it comes to your attention there are discourse threads discussing something relevant to a PR or issue, put all links in the top, original issue or PR post (the one with the copyright information, etc) so that it's all collected in one place. I think doing the same thing on discourse (adding all relevant github links to the original post) is probably also good.

Thoughts on this rule?

As far as what-goes-where, I see discourse more as a forum for discussion that we tend to assume has a larger audience than a github PR. So open-ended design discussions, community-wide alerts, that kind of stuff can go there. I'd code-specific stuff in Github, as Daniel was saying, as long as it isn't an issue that requires a wider or longer discussion.

I think it'd be good to write this up and put it on a wiki somewhere, but I don't currently understand how our wikis are organized and I'm not really sure new people would understand even which repo to look at for wikis, nevermind which ones are relevant to them. A separate problem, haha.
